File: app/src/main/java/com/worldventures/dreamtrips/wallet/ui/settings/general/firmware/reset/poweron/impl/ForceUpdatePowerOnScreenImpl.java
package com.worldventures.dreamtrips.wallet.ui.settings.general.firmware.reset.poweron.impl;


import android.animation.AnimatorSet;
import android.support.annotation.NonNull;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;

import com.afollestad.materialdialogs.MaterialDialog;
import com.worldventures.dreamtrips.R;
import com.worldventures.dreamtrips.wallet.ui.common.base.WalletBaseController;
import com.worldventures.dreamtrips.wallet.ui.settings.general.firmware.reset.poweron.ForceUpdatePowerOnPresenter;
import com.worldventures.dreamtrips.wallet.ui.settings.general.firmware.reset.poweron.ForceUpdatePowerOnScreen;
import com.worldventures.dreamtrips.wallet.ui.widget.WizardVideoView;

import java.util.Arrays;

import javax.inject.Inject;

import static android.animation.ObjectAnimator.ofFloat;

public class ForceUpdatePowerOnScreenImpl extends WalletBaseController<ForceUpdatePowerOnScreen, ForceUpdatePowerOnPresenter> implements ForceUpdatePowerOnScreen {

   private static final int SHOW_SOAR_TITLE_DELAY = 1000;
   private static final int CARD_FADE_IN_DELAY = 300;
   private static final int COMMON_FADE_IN_DELAY = 250;

   private TextView walletWizardSplashTitle;
   private Button actionBtn;
   private WizardVideoView wizardVideoView;

   @Inject ForceUpdatePowerOnPresenter presenter;

   @Override
   protected void onFinishInflate(View view) {
      super.onFinishInflate(view);
      final Toolbar toolbar = view.findViewById(R.id.toolbar);
      toolbar.setNavigationOnClickListener(v -> getPresenter().onBack());
      walletWizardSplashTitle = view.findViewById(R.id.wallet_wizard_power_on_title);
      actionBtn = view.findViewById(R.id.wallet_wizard_power_on_btn);
      wizardVideoView = view.findViewById(R.id.wizard_video_view);
      hideAllView();
      wizardVideoView.setVideoSource(R.raw.wallet_anim_power_on_sc);
   }

   @Override
   public View inflateView(LayoutInflater layoutInflater, ViewGroup viewGroup) {
      return layoutInflater.inflate(R.layout.screen_wallet_force_fw_update_power_on, viewGroup, false);
   }

   @Override
   public boolean supportConnectionStatusLabel() {
      return false;
   }

   @Override
   public boolean supportHttpConnectionStatusLabel() {
      return false;
   }

   @Override
   protected void onAttach(@NonNull View view) {
      super.onAttach(view);
      getView().postDelayed(this::startSoarAnimation, SHOW_SOAR_TITLE_DELAY);
   }

   @Override
   public ForceUpdatePowerOnPresenter getPresenter() {
      return presenter;
   }

   private void hideAllView() {
      for (View view : Arrays.asList(actionBtn, walletWizardSplashTitle, wizardVideoView)) {
         view.setAlpha(0);
      }
   }

   private void startSoarAnimation() {
      AnimatorSet mainAnimation = new AnimatorSet();
      mainAnimation
            .play(ofFloat(actionBtn, View.ALPHA, 1).setDuration(COMMON_FADE_IN_DELAY))
            .with(ofFloat(walletWizardSplashTitle, View.ALPHA, 1).setDuration(COMMON_FADE_IN_DELAY))
            .after(ofFloat(wizardVideoView, View.ALPHA, 1).setDuration(CARD_FADE_IN_DELAY));

      mainAnimation.start();
   }


   @Override
   public void setButtonAction(View.OnClickListener onClickListener) {
      actionBtn.setOnClickListener(onClickListener);
   }

   @Override
   public void showDialogEnableBleAndInternet() {
      new MaterialDialog.Builder(getContext())
            .content(R.string.wallet_firmware_pre_installation_bluetooth)
            .positiveText(R.string.ok)
            .show();
   }
}

File: wallet/src/main/java/com/worldventures/wallet/ui/settings/general/firmware/reset/poweron/impl/ForceUpdatePowerOnScreenImpl.kt
package com.worldventures.wallet.ui.settings.general.firmware.reset.poweron.impl

import android.animation.AnimatorSet
import android.animation.ObjectAnimator.ofFloat
import android.support.v7.widget.Toolbar
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.TextView
import com.afollestad.materialdialogs.MaterialDialog
import com.worldventures.wallet.R
import com.worldventures.wallet.ui.common.base.WalletBaseController
import com.worldventures.wallet.ui.settings.general.firmware.reset.poweron.ForceUpdatePowerOnPresenter
import com.worldventures.wallet.ui.settings.general.firmware.reset.poweron.ForceUpdatePowerOnScreen
import com.worldventures.wallet.ui.widget.WizardVideoView
import javax.inject.Inject

class ForceUpdatePowerOnScreenImpl : WalletBaseController<ForceUpdatePowerOnScreen, ForceUpdatePowerOnPresenter>(), ForceUpdatePowerOnScreen {

   private lateinit var walletWizardSplashTitle: TextView
   private lateinit var nexrButton: Button
   private lateinit var wizardVideoView: WizardVideoView

   @Inject lateinit var screenPresenter: ForceUpdatePowerOnPresenter

   override fun onFinishInflate(view: View) {
      super.onFinishInflate(view)
      val toolbar = view.findViewById<Toolbar>(R.id.toolbar)
      toolbar.setNavigationOnClickListener { presenter.onBack() }
      walletWizardSplashTitle = view.findViewById(R.id.wallet_wizard_power_on_title)
      nexrButton = view.findViewById(R.id.wallet_wizard_power_on_btn)
      wizardVideoView = view.findViewById(R.id.wizard_video_view)
      hideAllView()
      wizardVideoView.setVideoSource(R.raw.wallet_anim_power_on_sc)

      nexrButton.setOnClickListener { presenter.goNext() }
   }

   override fun inflateView(layoutInflater: LayoutInflater, viewGroup: ViewGroup) =
         layoutInflater.inflate(R.layout.screen_wallet_force_fw_update_power_on, viewGroup, false)!!

   override fun supportConnectionStatusLabel() = false

   override fun supportHttpConnectionStatusLabel() = false

   override fun onAttach(view: View) {
      super.onAttach(view)
      view.postDelayed({ this.startSoarAnimation() }, SHOW_SOAR_TITLE_DELAY)
   }

   override fun getPresenter() = screenPresenter

   private fun hideAllView() {
      listOf(nexrButton, walletWizardSplashTitle, wizardVideoView).forEach { it.alpha = 0f }
   }

   private fun startSoarAnimation() {
      val mainAnimation = AnimatorSet()
      mainAnimation
            .play(ofFloat<View>(nexrButton, View.ALPHA, 1f).setDuration(COMMON_FADE_IN_DELAY))
            .with(ofFloat<View>(walletWizardSplashTitle, View.ALPHA, 1f).setDuration(COMMON_FADE_IN_DELAY))
            .after(ofFloat<View>(wizardVideoView, View.ALPHA, 1f).setDuration(CARD_FADE_IN_DELAY))

      mainAnimation.start()
   }

   override fun showDialogEnableBleAndInternet() {
      MaterialDialog.Builder(context)
            .content(R.string.wallet_firmware_pre_installation_bluetooth)
            .positiveText(R.string.wallet_ok)
            .show()
   }

   companion object {

      private val SHOW_SOAR_TITLE_DELAY = 1000L
      private val CARD_FADE_IN_DELAY = 300L
      private val COMMON_FADE_IN_DELAY = 250L
   }
}
